File: tst/collide_pct.c
#include "pico.h"
#include "../check.h"

int main() {
    pico_init(1);

    Pico_Rel_Dim phy = { '!', {200,200}, NULL };
    Pico_Rel_Dim log = { '!', { 20, 20}, NULL };
    pico_set_view(NULL, -1, -1, &phy, NULL, &log, NULL, NULL, NULL);

    Pico_Rel_Rect r = { '%', {0.5,0.5, 0.5,0.5}, PICO_ANCHOR_C, NULL };

    puts("pos_vs_rect");
    for (float y=-1; y<=10; y+=1) {
        for (float x=-1; x<=10; x+=1) {
            pico_output_clear();
            pico_set_color_draw((Pico_Color){255,255,255});
            pico_output_draw_rect(&r);

            Pico_Rel_Pos p = { '%', {x/10.0,y/10.0}, PICO_ANCHOR_C, &r };
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_pixel(&p);

            int in = pico_vs_pos_rect(&p, &r);
            puts(in ? "in" : "out");

            if (x==-1 && y==-1) {
                assert(!in);
                _pico_check("collide_pct-01");
            }
            if (x==0 && y==0) {
                assert(in);
                _pico_check("collide_pct-02");
            }
            if (x==9 && y==9) {
                assert(in);
                _pico_check("collide_pct-03");
            }
            if (x==10 && y==10) {
                assert(!in);
                _pico_check("collide_pct-04");
            }

            pico_input_delay(10);
        }
    }

    puts("rect_vs_rect - same anchor");
    for (float y=-25; y<=125; y+=10) {
        for (float x=-25; x<=125; x+=10) {
            pico_output_clear();
            pico_set_color_draw((Pico_Color){255,255,255});
            pico_output_draw_rect(&r);

            Pico_Rel_Rect r2 = { '%', {x/100.0,y/100.0, 0.5,0.5}, PICO_ANCHOR_C, &r };
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_rect(&r2);

            int in = pico_vs_rect_rect(&r2, &r);
            puts(in ? "overlap" : "naw");

            if (x==-25 && y==-25) {
                assert(!in);
                _pico_check("collide_pct-05");
            }
            if (x==-15 && y==-15) {
                assert(in);
                _pico_check("collide_pct-06");
            }
            if (x==115 && y==115) {
                assert(in);
                _pico_check("collide_pct-07");
            }
            if (x==125 && y==125) {
                assert(!in);
                _pico_check("collide_pct-08");
            }

            pico_input_delay(10);
        }
    }

    puts("rect_vs_rect - diff anchor");
    for (float y=-5; y<=10; y+=1) {
        for (float x=-5; x<=10; x+=1) {
            pico_output_clear();
            pico_set_color_draw((Pico_Color){255,255,255});
            pico_output_draw_rect(&r);

            Pico_Rel_Rect r2 = { '%', {x/10.0,y/10.0, 0.5,0.5}, PICO_ANCHOR_NW, &r };
            pico_set_color_draw((Pico_Color){255,0,0});
            pico_output_draw_rect(&r2);

            int in = pico_vs_rect_rect(&r2, &r);
            puts(in ? "overlap" : "naw");

            if (x==-5 && y==-5) {
                assert(!in);
                _pico_check("collide_pct-09");
            }
            if (x==-4 && y==-4) {
                assert(in);
                _pico_check("collide_pct-10");
            }
            if (x==9 && y==9) {
                assert(in);
                _pico_check("collide_pct-11");
            }
            if (x==10 && y==10) {
                assert(!in);
                _pico_check("collide_pct-12");
            }

            pico_input_delay(10);
        }
    }

    pico_init(0);
    return 0;
}

File: lua/tst/collide_pct.lua
require 'pico.check'

pico.init(true)

local phy = {'!', w=200, h=200}
local log = {'!', w=20, h=20}
pico.set.window { dim=phy }
pico.set.view { dim=log }

local r = {'%', x=0.5, y=0.5, w=0.5, h=0.5}

print("pos_vs_rect")
for y = -1, 10 do
    for x = -1, 10 do
        pico.output.clear()
        pico.set.color.draw(255, 255, 255)
        pico.output.draw.rect(r)

        local p = {'%', x=x/10.0, y=y/10.0, up=r}
        pico.set.color.draw(255, 0, 0)
        pico.output.draw.pixel(p)

        local on = pico.vs.pos_rect(p, r)
        print(on and "in" or "out")

        if x==-1 and y==-1 then
            assert(not on)
            pico.check("collide_pct-01")
        end
        if x==0 and y==0 then
            assert(on)
            pico.check("collide_pct-02")
        end
        if x==9 and y==9 then
            assert(on)
            pico.check("collide_pct-03")
        end
        if x==10 and y==10 then
            assert(not on)
            pico.check("collide_pct-04")
        end

        pico.input.delay(10)
    end
end

print("rect_vs_rect - same anchor")
for y = -25, 125, 10 do
    for x = -25, 125, 10 do
        pico.output.clear()
        pico.set.color.draw(255, 255, 255)
        pico.output.draw.rect(r)

        local r2 = {'%', x=x/100.0, y=y/100.0, w=0.5, h=0.5, up=r}
        pico.set.color.draw(255, 0, 0)
        pico.output.draw.rect(r2)

        local overlap = pico.vs.rect_rect(r2, r)
        print(overlap and "overlap" or "naw")

        if x==-25 and y==-25 then
            assert(not overlap)
            pico.check("collide_pct-05")
        end
        if x==-15 and y==-15 then
            assert(overlap)
            pico.check("collide_pct-06")
        end
        if x==115 and y==115 then
            assert(overlap)
            pico.check("collide_pct-07")
        end
        if x==125 and y==125 then
            assert(not overlap)
            pico.check("collide_pct-08")
        end

        pico.input.delay(10)
    end
end

print("rect_vs_rect - diff anchor")
for y = -5, 10 do
    for x = -5, 10 do
        pico.output.clear()
        pico.set.color.draw(255, 255, 255)
        pico.output.draw.rect(r)

        local r2 = {'%', x=x/10.0, y=y/10.0, w=0.5, h=0.5, anc='NW', up=r}
        pico.set.color.draw(255, 0, 0)
        pico.output.draw.rect(r2)

        local overlap = pico.vs.rect_rect(r2, r)
        print(overlap and "overlap" or "naw")

        if x==-5 and y==-5 then
            assert(not overlap)
            pico.check("collide_pct-09")
        end
        if x==-4 and y==-4 then
            assert(overlap)
            pico.check("collide_pct-10")
        end
        if x==9 and y==9 then
            assert(overlap)
            pico.check("collide_pct-11")
        end
        if x==10 and y==10 then
            assert(not overlap)
            pico.check("collide_pct-12")
        end

        pico.input.delay(10)
    end
end

pico.init(false)
